packages/smithy-client/src/parse-utils.ts

```diff
@@ -262,19 +255,28 @@ export const strictParseFloat = strictParseDouble;
  * @returns The value as a number, or undefined if it's null/undefined.
  */
 export const strictParseFloat32 = (value: string | number): number | undefined => {
-  if (value === "NaN") {
-    return NaN;
-  }
   if (typeof value == "string") {
-    const parsed: number = parseFloat(value);
-    if (Number.isNaN(parsed)) {
-      throw new TypeError(`Expected real number, got implicit NaN`);
-    }
-    return expectFloat32(parsed);
+    return expectFloat32(parseNumber(value));
   }
   return expectFloat32(value);
 };
 
+// This regex matches JSON-style numbers. In short:
+// * The integral may start with a negative sign, but not a positive one
+// * No leading 0 on the integral unless it's immediately followed by a '.'
+// * Exponent indicated by a case-insensitive 'E' optionally followed by a
+//   positive/negative sign and some number of digits.
+// It also matches both positive and negative infinity as well and explicit NaN.
+const NUMBER_REGEX = /(-?((0(\.\d+)?)|(([1-9]\d*(\.\d)?)\d*)((e|E)(\+|-)?\d+)?))|(-?Infinity)|(NaN)/g;
```

Comment:
Saw `/-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?/` on stack overflow, which seems like a pretty compact way to write it, though I think this is equivalent.

Reply:
Yeah I didn't really spend any time regex-golfing, I probably should have just googled it in the first place